android/src/main/java/com/marigold/rnsdk/RNEngageBySailthruModule.kt

```diff
@@ -0,0 +1,347 @@
+            convertedTags = ArrayList()
+            for (i in 0 until tags.size()) {
+                convertedTags.add(tags.getString(i))
+            }
+        }
+        engage.trackPageview(uri, convertedTags, object : EngageBySailthru.TrackHandler {
+            override fun onSuccess() {
+                promise.resolve(true)
+            }
+
+            override fun onFailure(error: Error) {
+                promise.reject(RNMarigoldModule.ERROR_CODE_TRACKING, error.message)
+            }
+        })
+    }
+
+    @ReactMethod
+    fun trackImpression(sectionId: String, urls: ReadableArray?, promise: Promise) {
+        var convertedUrls: List<URI>? = null
+        if (urls != null) {
+            try {
+                convertedUrls = ArrayList()
+                for (i in 0 until urls.size()) {
+                    convertedUrls.add(URI(urls.getString(i)))
+                }
+            } catch (e: URISyntaxException) {
+                promise.reject(RNMarigoldModule.ERROR_CODE_TRACKING, e.message)
+                return
+            }
+        }
+        engage.trackImpression(sectionId, convertedUrls, object : EngageBySailthru.TrackHandler {
+            override fun onSuccess() {
+                promise.resolve(true)
+            }
+
+            override fun onFailure(error: Error) {
+                promise.reject(RNMarigoldModule.ERROR_CODE_TRACKING, error.message)
+            }
+        })
+    }
+
+    @ReactMethod
+    fun setProfileVars(vars: ReadableMap, promise: Promise) {
+        val varsJson = try {
+            jsonConverter.convertMapToJson(vars)
+        } catch (e: JSONException) {
+            promise.reject(RNMarigoldModule.ERROR_CODE_VARS, e.message)
+            return
+        }
+        engage.setProfileVars(varsJson, object : Marigold.MarigoldHandler<Void?> {
+            override fun onSuccess(value: Void?) {
+                promise.resolve(true)
+            }
+
+            override fun onFailure(error: Error) {
+                promise.reject(RNMarigoldModule.ERROR_CODE_VARS, error.message)
+            }
+        })
+    }
+
+    @ReactMethod
+    fun getProfileVars(promise: Promise) {
+        engage.getProfileVars(object : Marigold.MarigoldHandler<JSONObject?> {
+            override fun onSuccess(value: JSONObject?) {
+                try {
+                    val vars = value?.let { jsonConverter.convertJsonToMap(it) }
+                    promise.resolve(vars)
+                } catch (e: JSONException) {
+                    promise.reject(RNMarigoldModule.ERROR_CODE_VARS, e.message)
+                }
+            }
+
+            override fun onFailure(error: Error) {
+                promise.reject(RNMarigoldModule.ERROR_CODE_VARS, error.message)
+            }
+        })
+    }
+
+    @ReactMethod
+    fun logPurchase(purchaseMap: ReadableMap, promise: Promise) {
+        val purchase = try {
+            getPurchaseInstance(purchaseMap)
+        } catch (e: JSONException) {
```

Comment:
All these different exception handlers implement the same handling (in both logPurchase and logAbandonedCart). We could aim for something a little less verbose, like for instance:
```
fun getPurchaseInstance(purchaseMap: ReadableMap, promise: Promise): Purchase? = try {
    val purchaseJson = jsonConverter.convertMapToJson(purchaseMap, false)
    val purchaseConstructor = Purchase::class.java.getDeclaredConstructor(JSONObject::class.java)
    purchaseConstructor.isAccessible = true
    purchaseConstructor.newInstance(purchaseJson)
} catch(e: Exception) {
    promise.reject(RNMarigoldModule.ERROR_CODE_PURCHASE, e.message)
    null
}
```
so then in the methods you could just call:
```
val purchase = getPurchaseInstance(purchaseMap, promise) ?: return
```